// @cityofzion/neon-core/lib/wallet/Claims.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = __importDefault(require("util"));
const ClaimItem_1 = __importDefault(require("./components/ClaimItem"));
/**
 * Claims object used for claiming GAS.
 */
class Claims {
    constructor(config = {}) {
        this.address = config.address || "";
        this.net = config.net || "NoNet";
        this.claims = config.claims ? config.claims.map(c => new ClaimItem_1.default(c)) : [];
    }
    get [Symbol.toStringTag]() {
        return "Claims";
    }
    [util_1.default.inspect.custom]() {
        const claimsDump = this.claims.map(c => {
            return `${c.txid} <${c.index}>: ${c.claim.toString()}`;
        });
        return `[Claims(${this.net}): ${this.address}]\n${JSON.stringify(claimsDump, null, 2)}`;
    }
    export() {
        return {
            address: this.address,
            net: this.net,
            claims: this.claims.map(c => c.export())
        };
    }
    /**
     * Returns new Claims object that contains part of the total claims starting at start, ending at end.
     */
    slice(start, end) {
        return new Claims({
            address: this.address,
            net: this.net,
            claims: this.claims.slice(start, end)
        });
    }
}
exports.Claims = Claims;
exports.default = Claims;
